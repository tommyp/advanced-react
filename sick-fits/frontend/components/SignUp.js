/* eslint-disable no-underscore-dangle */
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Router from 'next/dist/client/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
// import { CURRENT_USER_QUERY } from './User';
import DisplayError from './ErrorMessage';

const SIGNUP_MUTATION = gql`
mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
  createUser(data: {
    email: $email,
    name: $name,
    password: $password,
  }) {
    id
    email
    name
  }
}
`;

export default function SignUp() {
  const { inputs, handleChange, resetForm } = useForm({
    name: '',
    email: '',
    password: '',

  });

  const [signUp, { data, error, loading }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
    // refetchQueries: [{
    //   query: CURRENT_USER_QUERY,
    // }],
  });

  const handleSubmit = async (e) => {
    e.preventDefault();

    await signUp().catch(console.error);
    resetForm();
  };

  return (
    <>
      <Form onSubmit={handleSubmit} method="POST">
        <DisplayError error={error} />
        <h2>Sign up for an account</h2>
        <fieldset>
          {data?.createUser && (
          <p>
            Signed up with
            {' '}
            {data.createUser.email}
            {' '}
            - please go ahead and sign in
          </p>
          )}
          <label htmlFor="name">
            Name
            <input type="text" name="name" id="name" onChange={handleChange} placeholder="Your name " autoComplete="name" value={inputs.name} />
          </label>
          <label htmlFor="email">
            Email
            <input type="email" name="email" id="email" onChange={handleChange} placeholder="Your Email Addres" autoComplete="email" value={inputs.email} />
          </label>
          <label htmlFor="password">
            Password
            <input type="password" name="password" id="password" onChange={handleChange} placeholder="Your password" value={inputs.password} />
          </label>
          <button type="submit">Sign in</button>
        </fieldset>
      </Form>
    </>
  );
}
