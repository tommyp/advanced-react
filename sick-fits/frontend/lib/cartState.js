/* eslint-disable import/prefer-default-export */
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider!
  // We will store data (state) and functionality
  // (updaters) in here and anyone can access it via the consumer

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  return (
    <LocalStateProvider value={{ cartOpen, toggleCart }}>
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
