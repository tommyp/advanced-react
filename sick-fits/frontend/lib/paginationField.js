/* eslint-disable no-plusplus */
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will take care of everything
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });

      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      if (items.length && items.length !== first && page === pages) {
        return items;
      }

      if (items.length !== first) {
        // we don't have any items, we must go to the network
        return false;
      }

      // if there are items, just return them from the cache and we don't need to go to the network

      // if there are items AND
      // there aren't enough items to satisfy how many were request
      // AND if we're on the last page

      if (items.length) {
        return items;
      }

      return false;
      // first thing it does it asks the read function for those items

      // we can either do one of two things

      // 1st: return the items from the cache

      // 2nd: return false from here, which causes a network request
    },
    merge(existing, incoming, { args }) {
      // this runs when the apollo client comes back from the network with our products
      const { skip, first } = args;

      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }

      return merged;
    },
  };
}
