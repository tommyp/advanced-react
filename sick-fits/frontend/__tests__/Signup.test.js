import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
import { fakeUser } from '../lib/testUtils';

const user = fakeUser();
const password = 'password';

const mocks = [
  // mutation mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: user.name,
        email: user.email,
        password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: user.id,
          email: user.email,
          name: user.name,
        },
      },
    },
  },
];

describe('<SignUp/>', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('calls the mutation properly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>,
    );

    // type into boxes
    await userEvent.type(
      screen.getByLabelText('Name'),
      user.name,
    );

    await userEvent.type(
      screen.getByLabelText('Email'),
      user.email,
    );

    await userEvent.type(
      screen.getByLabelText('Password'),
      password,
    );

    await userEvent.click(
      screen.getByText('Sign up'),
    );

    await screen.findByText(`Signed up with ${user.email} - please go ahead and sign in`);
  });
});
