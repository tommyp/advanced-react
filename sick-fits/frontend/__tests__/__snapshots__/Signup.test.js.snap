// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
    method="POST"
  >
    <h2>
      Sign up for an account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          id="name"
          name="name"
          placeholder="Your name "
          type="text"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          id="email"
          name="email"
          placeholder="Your Email Addres"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          id="password"
          name="password"
          placeholder="Your password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign up
      </button>
    </fieldset>
  </form>
</div>
`;
